File: java/v1/src/main/java/com/api/v1/book/domain/Book.java
package com.api.v1.book.domain;

import com.api.v1.book.dtos.BookRequestDto;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.ZonedDateTime;
import java.util.UUID;

@Document(collection = "v1_books")
public class Book {

    @Id
    private UUID id;

    @Field
    private String title;

    @Field
    private String subtitle;

    @Field
    private String isbn;

    @Field
    private String author;

    @Field
    private String field;

    @Field
    private int numberOfPages;

    @Field
    private int version;

    @Field
    private String addedAt;

    @Field
    private String updatedAt;

    public Book() {}

    public Book(
            UUID id,
            String title,
            String subtitle,
            String isbn,
            String author,
            String field,
            int numberOfPages,
            int version,
            String addedAt
    ) {
        this.id = id;
        this.title = title;
        this.subtitle = subtitle;
        this.isbn = isbn;
        this.author = author;
        this.field = field;
        this.numberOfPages = numberOfPages;
        this.version = version;
        this.addedAt = addedAt;
    }

    public String getFullTitle() {
        if (subtitle.isEmpty()) return title;
        return "%s: %s".formatted(title, subtitle);
    }

    public void update(BookRequestDto request) {
        this.title = request.title();
        this.subtitle = request.subtitle();
        this.author = request.author();
        this.field = request.field();
        this.numberOfPages = request.numberOfPages();
        this.version = request.version();
        this.updatedAt = ZonedDateTime.now().toString();
    }

    public UUID getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getSubtitle() {
        return subtitle;
    }

    public String getIsbn() {
        return isbn;
    }

    public String getAuthor() {
        return author;
    }

    public String getField() {
        return field;
    }

    public int getNumberOfPages() {
        return numberOfPages;
    }

    public int getVersion() {
        return version;
    }

    public String getAddedAt() {
        return addedAt;
    }

    public String getUpdatedAt() {
        return updatedAt;
    }

}

File: koltlin/v1/src/main/kotlin/com/api/v1/book/domain/Book.kt
package com.api.v1.book.domain

import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.Field

import java.util.UUID;

@Document(collection = "v1_books")
class Book {

    @Id
    val id: UUID

    @Field
    var title: String

    @Field
    var subtitle: String

    @Field
    var isbn: String

    @Field
    var publisher: String

    @Field
    var version: Int

    @Field
    var numberOfPages: Int

    @Field
    var author: String

    @Field
    var field: String

    @Field
    val addedAt: String

    @Field
    val updatedAt: String = ""

    constructor(
        id: UUID,
        title: String,
        subtitle: String,
        isbn: String,
        publisher: String,
        version: Int,
        numberOfPages: Int,
        author: String,
        field: String,
        addedAt: String
    ) {
        this.id = id
        this.title = title
        this.subtitle = subtitle
        this.isbn = isbn
        this.publisher = publisher
        this.version = version
        this.numberOfPages = numberOfPages
        this.author = author
        this.field = field
        this.addedAt = addedAt
    }

}